### DISCUSSION TITLE
Qual a diferença entre biblioteca e framework?

### DISCUSSION THREAD

**Discussion by luisfeliperm** [frontendbr/forum#1281] [Perguntas e Respostas]
Ainda não consigo entender qual a diferença entre biblioteca e framework.

font-awesome por exemplo, é um biblioteca ou framework?

**Comment by fxcosta** [comment]
Eu vi uma discussão sobre o assunto há um tempo e gostei da seguinte definição:

- Uma lib (biblioteca) te fornece uma API para trabalhar com algo, como um serviço. Você geralmente usa ela com várias outras APIs distintas ou seja, mesclar várias libs sem muitos problemas e tem, geralmente, um alto grau de flexibilidade.

- Um framework tbm te fornece uma API para trabalha mas ele te fornece mais que isso: ele te impõem uma arquitetura. Essa é mais complexa de gerenciar e de mesclar com várias outras coisas diferentes.

Um exemplo? Lodash é uma biblioteca, ela te fornece uma série de APIs bacanas pra você lidar com estruturas complexas no Javascript. Um framework seria o Angular. Ele além de te entregar uma API, te entrega uma arquitetura inteira e completa (DI, router, templating, service layers...)

Porém, nem todo framework é um apunhado de ferramentas inteiras. De novo, um framework é aquele código que te "obriga" a trabalhar naquela arquitetura e ir por um caminho diferente meio que não faz sentido. Um outro exemplo de framework são as ferramentas de testes unitários. Eles te fornecem uma API para asserções mas, novamente, te obriga a trabalhar em cima de uma arquitetura inteira (configurações, convenções, métodos, etc).

**Comment by fxcosta** [comment]
Me desculpe, acabei não respondendo sobre o font-awesome que acredito ser o principal. Usando a ideia que defendi, seria uma lib.

**Comment by luisfeliperm** [comment]
Fxcosta vlw, sua definição foi bem resumida e agora eu entendi, obrigado

**Comment by felipefialho** [comment]
@luisfeliperm Editei o titulo e texto para ficar mais generico :)

**Comment by reinaldovale** [comment]
Framework: (normalmente) grande e engessado.
Biblioteca: (normalmente) pequena e flexível, muitas vezes, intercambiáveis.

**Comment by klarkc** [comment]
Apenas para acrescentar, existem os chamados frameworks progressivos, que partem de uma atomicidade tal qual uma biblioteca, mas oferecem uma gama de extensões e por vezes até um toolkit que lhe permite ir incrementando e se aproximando da arquitetura própria do framework. Exemplo: Vuejs

**Comment by didiraja** [comment]
Dada a explicação do @fxcosta, que eu achei correta, o React seria um framework mas todo mundo trata, e a própria documentação diz, que é uma Lib. E aí?

**Comment by klarkc** [comment]
Será q da pra chamar React + Redux de framework e React only uma lib? Pessoalmente eu acho q o react caia melhor na definição de framework progressivo, já que ele impõe uma forma de se escrever componentes usando o virtual dom, tal qual o Vuejs.

**Comment by fxcosta** [comment]
@didiraja @klarkc é interessante essa discussão porque ela beira a filosofia. react pra mim é uma lib. react + redux pra mim são duas libs, até porque redux vive sem o react e vice-versa.

se vc me entregar o react + react-router + uma série de coisas, aí talvez tenhamos um ecossistema que beira um framework mas perceba a diferença entre o react-router com o router do angular? não é só porque é separado, é porque o uso do react-router é opcional, vc pode usar outra solução. React-router seria uma extensão de uma lib, mas não uma lib e nem um framework.

O que o @klarkc chama de frameworks progressivos eu conheço melhor com o uso de micro-frameworks. No mundo do PHP isso é mais visto, não conheço nenhum micro-framework em js. Mas no PHP a ideia de micro-frameworks é te oferecer uma pequena estrutura, uma arquitetura de lego. Você vai montando suas peças e crescendo o app da forma como precisar ao invés de ter uma base de código enorme que talvez você nem use ou nem curta.

E se eu não curtir o roteador do angular? eu posso trocar? (desconheço mas acredito q não, se tiver errado me falem)
e o roteador do react? eu consigo usar algo além dele ou usar sem ele?

**Comment by klarkc** [comment]
Em JS, diferentemente de muitas linguagens, tudo é muito modular e desacoplado, então acho que qualquer framework é "plugável" com qualquer lib ou ferramenta, no final tudo é código e a diferença entre usar uma lib e usar um framework é apenas semântica e pouco prática.

Uma definição que talvez caiba melhor é que framework JS é um produto, uma caixa de ferramentas de módulos, no sentido que você dispõe dessas ferramentas standard, que são sugeridas para se ter maior eficiência ao escrever código, mas todas são opcionais.
